From 58fecf85ce725722ddf1da0ca23a1e3435c058f3 Mon Sep 17 00:00:00 2001
From: Abin Joseph <abin.joseph@amd.com>
Date: Fri, 11 Apr 2025 17:10:53 +0530
Subject: [PATCH 105/110] net: axienet: Check if auto negotiation is enabled
 before using it

XXV PCS feature implementation currently relies on AN status register.
But this feature is optional and inaccessible when AN/LT is disabled in
the IP at the synthesis time. Use the new auto negotiation property to
check for this feature. Default to AN off when the property is not
present to ensure backward compatibility for existing users without AN
in their designs.

Signed-off-by: Abin Joseph <abin.joseph@amd.com>
Reviewed-by: Harini Katakam <harini.katakam@amd.com>
Message-ID: <20250411114053.1240266-3-abin.joseph@amd.com>
---
 drivers/net/ethernet/xilinx/xilinx_axienet.h  |  2 ++
 .../net/ethernet/xilinx/xilinx_axienet_main.c | 30 ++++++++++++-------
 2 files changed, 21 insertions(+), 11 deletions(-)

diff --git a/drivers/net/ethernet/xilinx/xilinx_axienet.h b/drivers/net/ethernet/xilinx/xilinx_axienet.h
index 0b6ab7abc951d..f9a7e79293aae 100644
--- a/drivers/net/ethernet/xilinx/xilinx_axienet.h
+++ b/drivers/net/ethernet/xilinx/xilinx_axienet.h
@@ -1016,6 +1016,7 @@ struct skbuf_dma_descriptor {
  * @gt_mode_narrow: true if GT is configured to operate in Narrow mode, false for Wide mode.
  * @mrmac_stream_dwidth: MRMAC AXI4-Stream data width (bits).
  * @switch_lock: Spinlock for switchable IP.
+ * @auto_neg: true if auto neg property is enabled in the IP.
  * @eoe_regs: Ethernet offload IP base address.
  * @eoe_connected: Tells whether ethernet offload IP is connected to Ethernet IP.
  * @eoe_features: EOE IP supported configuration.
@@ -1135,6 +1136,7 @@ struct axienet_local {
 	bool gt_mode_narrow;
 	int mrmac_stream_dwidth;
 	spinlock_t switch_lock;	/* To protect Link training programming from multiple context */
+	bool auto_neg;
 	void __iomem *eoe_regs;
 	bool eoe_connected;
 	u32 eoe_features;
diff --git a/drivers/net/ethernet/xilinx/xilinx_axienet_main.c b/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
index cb1c25bcae868..5576020391b3c 100644
--- a/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
+++ b/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
@@ -4155,12 +4155,14 @@ static void axienet_pcs_get_state(struct phylink_pcs *pcs,
 			state->speed = SPEED_25000;
 
 		state->duplex = DUPLEX_FULL;
-		an_status = axienet_ior(lp, XXV_STAT_AN_STS_OFFSET);
-		tx_pause = an_status & XXV_TX_PAUSE_MASK;
-		rx_pause = an_status & XXV_RX_PAUSE_MASK;
+		if (lp->auto_neg) {
+			an_status = axienet_ior(lp, XXV_STAT_AN_STS_OFFSET);
+			tx_pause = an_status & XXV_TX_PAUSE_MASK;
+			rx_pause = an_status & XXV_RX_PAUSE_MASK;
 
-		state->pause = (tx_pause & MLO_PAUSE_TX) | (rx_pause & MLO_PAUSE_RX);
-		state->an_complete = an_status & XXV_AN_COMPLETE_MASK;
+			state->pause = (tx_pause & MLO_PAUSE_TX) | (rx_pause & MLO_PAUSE_RX);
+			state->an_complete = an_status & XXV_AN_COMPLETE_MASK;
+		}
 		state->link = 0;
 
 		gt_rst = readl_poll_timeout(lp->regs + XXV_STAT_GTWIZ_OFFSET,
@@ -4240,14 +4242,16 @@ static int axienet_pcs_config(struct phylink_pcs *pcs, unsigned int neg_mode,
 		}
 	}
 	if (lp->axienet_config->mactype == XAXIENET_10G_25G) {
-		u32 autoneg_complete;
+		if (lp->auto_neg) {
+			u32 autoneg_complete;
 
-		autoneg_complete = (axienet_ior(lp, XXV_STAT_AN_STS_OFFSET) &
-				    XXV_AN_COMPLETE_MASK);
+			autoneg_complete = (axienet_ior(lp, XXV_STAT_AN_STS_OFFSET) &
+					    XXV_AN_COMPLETE_MASK);
 
-		/* If auto-negotiation is not completed, restart auto-neg */
-		return (neg_mode == (unsigned int)PHYLINK_PCS_NEG_INBAND_ENABLED &&
-			autoneg_complete == 0);
+			/* If auto-negotiation is not completed, restart auto-neg */
+			return (neg_mode == (unsigned int)PHYLINK_PCS_NEG_INBAND_ENABLED &&
+				autoneg_complete == 0);
+		}
 	} else if (lp->axienet_config->mactype == XAXIENET_1G_10G_25G) {
 		bool an_enabled = false;
 
@@ -5027,6 +5031,10 @@ static int axienet_probe(struct platform_device *pdev)
 	lp->eth_hasptp = of_property_read_bool(pdev->dev.of_node,
 					       "xlnx,eth-hasptp");
 
+	if (lp->axienet_config->mactype == XAXIENET_10G_25G)
+		lp->auto_neg = of_property_read_bool(pdev->dev.of_node,
+						     "xlnx,has-auto-neg");
+
 	if (lp->axienet_config->mactype == XAXIENET_10G_25G  ||
 	    lp->axienet_config->mactype == XAXIENET_1G_10G_25G)
 		lp->xxv_ip_version = axienet_ior(lp, XXV_CONFIG_REVISION);
-- 
2.43.0

